rfd/0032-role-tester.md

```diff
@@ -0,0 +1,94 @@
+
+Fact | Example | Interpretation
+--- | --- | ---
+HasRole(user, role) | HasRole(jean, dev) | User 'jean' has role 'dev'
+NodeHasSSHLabel(node, label_key, label_value) | NodeHasSSHLabel(node-1, environment, staging) | SSH node 'node-1' has the label 'environment:staging'
+RoleAllowedSSHLabel(role, label_key, label_value) | RoleAllowedSSHLabel(dev, environment, staging) | Role 'dev' is allowed access to SSH nodes with label 'environment:staging'
+RoleDeniedSSHLabel(role, label_key, label_value) | RoleDeniedSSHLabel(bad, environment, production) | Role 'bad' is denied access to SSH nodes with label 'environment:production'
+RoleHasSSHLogin(role, login) | RoleHasSSHLogin(admin, root) | Role 'admin' can login as os user 'root' to SSH nodes
+
+These facts are then used within rules to answer questions. The facts themselves can also be queried, for example HasRole(jean, Role)? will return all the roles that user 'jean' has where the term 'Role' represents an arbitrary variable and 'jean' is a constant.
+
+One concern would be that a user can have multiple roles. The way Datalog works allows for this one to many mapping by just adding multiple facts, which in this case is multiple HasRole facts for the user. For example:
+
+```prolog
+HasRole(jean, dev).
+HasRole(jean, cloud).
+HasRole(jean, admin).
+```
+means that Jean has roles 'dev', 'cloud', and 'admin'. Similarly, this representation also applies to labels and logins for roles and nodes.
+
+### Rules
+Rules are sentences that allow us to infer new facts from existing ones. We can combine multiple rules to create new rules. The most important question for the role tester is whether a user can access a given node as an os user, and we will define this as HasAccess(User, Login, Node). Other rules will provide other contextually related facts that will help answer this overarching question.
+
+Rule | Logical interpretation
+--- | ---
+HasAllowedSSHLabel(Role, Node, Key, Value) :- RoleAllowedSSHLabel(Role, Key, Value), NodeHasSSHLabel(Node, Key, Value) | If the role has allow labels and the node has the same labels, then the role allows access to the node
+HasDeniedSSHLabel(Role, Node, Key, Value) :- RoleDeniedSSHLabel(Role, Key, Value), NodeHasSSHLabel(Node, Key, Value) | If the role has deny labels and the node has the same labels, then the role denies access to the node
+HasRolePath(User, Login, Node) :- HasRole(User, Role), HasAllowedSSHLabel(Role, Node, Key, Value), RoleHasSSHLogin(Role, Login) | If the user has a role and the role allows access to the node and the role has a login, then the user has a role that gives them access to the node as login
+HasDeniedRolePath(User, Node) :- HasRole(User, Role), HasDeniedSSHLabel(Role, Node, Key, Value) | If the user has a role and the role denies access to the node, then the user has a role that denies them access to the node
```

Comment:
Should take `Login` as a parameter too.

Reply:
Mentioned in the comment above, but this will need to be added since there are login denies.